ambry-utils/src/test/java/com.github.ambry.utils/UtilsTest.java

```diff
@@ -192,6 +194,75 @@ public void testReadBuffers() throws IOException {
     }
   }
 
+  @Test
+  public void testBlobDataShareMemory() throws Exception {
+    int blobSize = 1000;
+    // The first 8 bytes are the size of blob, the next 1000 bytes are the blob content, the next 8 bytes are the crc
+    // value, and we do this twice.
+    int bufferSize = (Long.SIZE / Byte.SIZE + blobSize + Long.SIZE / Byte.SIZE) * 2;
+    byte[] firstRandomBytes = new byte[blobSize];
+    byte[] secondRandomBytes = new byte[blobSize];
+    new Random().nextBytes(firstRandomBytes);
+    new Random().nextBytes(secondRandomBytes);
+
+    ByteBuffer buffer = ByteBuffer.allocate(bufferSize);
+    ByteBufferOutputStream bbos = new ByteBufferOutputStream(buffer);
+
+    // Fill the buffer
+    byte[] arrayToFill = firstRandomBytes;
+    while (arrayToFill != null) {
+      CrcOutputStream crcStream = new CrcOutputStream(bbos);
+      DataOutputStream dos = new DataOutputStream(crcStream);
+      dos.writeLong((long) blobSize);
+      dos.write(arrayToFill);
+      buffer.putLong(crcStream.getValue());
+      if (arrayToFill == firstRandomBytes) {
+        arrayToFill = secondRandomBytes;
+      } else {
+        arrayToFill = null;
+      }
+    }
+
+    buffer.flip();
+    byte[] expectedArray = firstRandomBytes;
+    while (expectedArray != null) {
+      CrcInputStream cis = new CrcInputStream(new ByteBufferDataInputStream(buffer));
+      DataInputStream dis = new DataInputStream(cis);
+      long dataSize = dis.readLong();
+      assertEquals((long) dataSize, blobSize);
+      ByteBufferInputStream obtained = Utils.getByteBufferInputStreamFromCRCInputStream(cis, (int) dataSize);
+      // Make sure these two ByteBuffers actually share the underlying memory.
+      assertEquals(getByteArrayFromByteBuffer(buffer), getByteArrayFromByteBuffer(obtained.getByteBuffer()));
```

Comment:
How does the below code ensure that the memory is shared, as opposed to the byte arrays being equals?

Reply:
Byte array being equal means these two ByteBuffers use the same byte array underlying. Here sharing memory means we are not creating a new byte array with the same bytes in it.